tests/postgresql/upgrade-with-extension.nix

```diff
@@ -0,0 +1,280 @@
+
+  testSetup = ''
+    # Make nix-build work inside the VM
+    machine.execute("mkdir -p /nix/var/nix/profiles/per-user/root/")
+    machine.execute("ln -s ${channel} /nix/var/nix/profiles/per-user/root/channels")
+
+    # Taken from upstream acme.nix
+    def switch_to(node, name, expect="succeed"):
+        # On first switch, this will create a symlink to the current system so that we can
+        # quickly switch between derivations
+        root_specs = "/tmp/specialisation"
+        node.execute(
+          f"test -e {root_specs}"
+          f" || ln -s $(readlink /run/current-system)/specialisation {root_specs}"
+        )
+
+        switcher_path = f"/run/current-system/specialisation/{name}/bin/switch-to-configuration"
+        rc, _ = node.execute(f"test -e '{switcher_path}'")
+        if rc > 0:
+            switcher_path = f"/tmp/specialisation/{name}/bin/switch-to-configuration"
+
+        if expect == "fail":
+          node.fail(f"{switcher_path} test")
+        else:
+          node.succeed(f"{switcher_path} test")
+
+    # helpers from the pg_anonymizer test in nixpkgs
+    def get_player_table_contents():
+        return [
+            x.split(',') for x in machine.succeed("sudo -u postgres psql -d anonymized --csv --command 'select * from player'").splitlines()[1:]
+        ]
+
+    def check_anonymized_row(row, id, original_name):
+        assert row[0] == id, f"Expected first row to have ID {id}, but got {row[0]}"
+        assert row[1] != original_name, f"Expected first row to have a name other than {original_name}"
+        assert not bool(row[2]), "Expected points to be NULL in first row"
+
+    def check_original_data(output):
+        assert output[0] == ['1','Foo','23'], f"Expected first row from player table to be 1,Foo,23; got {output[0]}"
+        assert output[1] == ['2','Bar','42'], f"Expected first row from player table to be 2,Bar,42; got {output[1]}"
+
+    def check_anonymized_rows(output):
+        check_anonymized_row(output[0], '1', 'Foo')
+        check_anonymized_row(output[1], '2', 'Bar')
+
+    machine.wait_for_unit("postgresql.service")
+    machine.wait_for_open_port(5432)
+
+    machine.succeed('sudo -u postgres -- sh ${dataTest}')
+
+    with subtest("Anonymize DB"):
+        check_original_data(get_player_table_contents())
+        machine.succeed("sudo -u postgres psql -d anonymized --command 'select anon.anonymize_database();'")
+  '';
+
+in {
+  name = "postgresql-upgrade-with-extensions";
+  testCases = {
+    manual = {
+      name = "manual";
+      nodes = {
+        machine = { ... }: {
+          imports = [
+            (testlib.fcConfig { net.fe = false; })
+          ];
+
+          flyingcircus.roles.postgresql12.enable = lib.mkDefault true;
+          services.postgresql = {
+            extraPlugins = ps: with ps; [ anonymizer ];
+            settings.shared_preload_libraries = lib.mkForce "auto_explain, pg_stat_statements, anon";
+          };
+
+          specialisation = {
```

Comment:
> This can be postponed though until when the autoupgrade tests are generally refactored.

Does this mean there was work on this already?
But yeah, I'd prefer to postpone this for now. I considered using the upgrade test for that, but splitting it off made it easier to iterate than having to invent a structure to embed these testcases into the existing ones on the fly.

Regarding modular tests: just in case you haven't seen it already, tests are also just modules which makes it relatively straightforward to write test helpers for a class of tests. See e.g. https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/nextcloud/default.nix.

Reply:
No, there was no proper work on refactoring so far, only on manually updating the versions in the test before. And that was when it became apparent that this just incurs a lot of manual labour.